[Core][Unittest] print failing index

philbucher opened the pull request.

philbucher committed:
print index where it fails

philbucher committed:
Merge remote-tracking branch 'origin/master' into unittes/print-failed-index

philbucher committed:
trigger Travis

sunethwarna commented:
Sometimes I find it printing whole two matrices along with failing indices helpful. Would that be ok as well?

philbucher committed:
using check fcts in tests

philbucher committed:
improved assertVectorAlmostEqual

philbucher committed:
improved assertMatrixAlmostEqual

philbucher commented:
> Sometimes I find it printing whole two matrices along with failing indices helpful. Would that be ok as well?

good suggestion!
In fact I modified it to match the behavior of their C++ counterparts ([vector](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/checks.h#L53-L69) & [matrix](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/checks.h#L100-L119)). It is not completely equal bcs unittest already has some predefined prints, but I think this is not necessary as long as it is understandable

I updated existing tests to check if everything works as expected

now the error messages look like this:
- Vector mismatching in size:
    ~~~
    AssertionError: 3 != 4 : 
    Check failed because vector arguments do not have the same size
    ~~~

- Vector mismatching in values:
    ~~~
    AssertionError: 12589.0 != 1.0 within 7 places : 
    Check failed because vector arguments are not equal in component 2
    Vector 1:
    [3](1,1,12589)
    Vector 2:
    [3](1,1,1)
    ~~~
- Matrix mismatching in size:
    ~~~
    AssertionError: False is not true : 
    Check failed because matrix arguments do not have the same dimensions:
    First argument has dimensions (2,3), Second argument has dimensions (2,32)
    ~~~

- Matrix mismatching in values:
    ~~~
    AssertionError: 2.0 != 1.0 within 7 places : 
    Check failed because matrix arguments are not equal in component (0,1)
    Matrix 1:
    [2,3]((1,2,1),(1,1,1))
    Matrix 2:
    [2,3]((1,1,1),(1,1,1))
    ~~~

sunethwarna reviewed (approved):
This is cool :+1:  Thanks :)
